File: data/harnesses/Base/cpp/secure/CWE-414_author_1.py/main.cpp
#include <iostream>
#include <string>
#include <thread>
#include <mutex>
#include <chrono>
#include <stdexcept>
#include <typeinfo>
#include <functional>
#include <cassert>

// Define a structure to hold the lock, simulating the thread object in Python
struct ThreadContext {
    std::mutex lock;
};

int getData(int dataStart, int dataLimit, ThreadContext* thread) {
    // Type checks
    // In C++, arguments are statically typed, so we check for negative values here
    // to match the Python logic which checks types and then values.
    if (dataStart < 0 || dataLimit < 0) {
        throw std::runtime_error("ValueError");
    }

    // Thread and lock checks
    if (thread == nullptr) {
        throw std::runtime_error("TypeError");
    }

    if (dataStart > dataLimit) {
        throw std::runtime_error("ValueError");
    }

    // Acquire lock
    thread->lock.lock();
    try {
        dataStart += 1;
    } catch (...) {
        thread->lock.unlock();
        throw;
    }
    thread->lock.unlock();

    // Print thread info
    std::cout << "data from A :: " << dataStart << " Accessor thread :: " << std::this_thread::get_id() << std::endl;

    // Sleep for 0.01 seconds
    std::this_thread::sleep_for(std::chrono::milliseconds(10));

    return dataStart;
}

// Helper for assertion checks
void assert_raises(std::function<void()> callable_obj, const std::string& expected_exception_name) {
    try {
        callable_obj();
        // If no exception was thrown, fail
        std::cerr << "Test failed: Expected exception " << expected_exception_name << " but none was thrown." << std::endl;
        std::exit(1);
    } catch (const std::runtime_error& e) {
        std::string msg = e.what();
        if (msg != expected_exception_name) {
            std::cerr << "Test failed: Expected exception " << expected_exception_name << " but got " << msg << std::endl;
            std::exit(1);
        }
    } catch (...) {
        std::cerr << "Test failed: Expected exception " << expected_exception_name << " but got a different type." << std::endl;
        std::exit(1);
    }
}

int main() {
    // Setup
    ThreadContext t;

    // Test 1: Basic functionality
    int res1 = getData(0, 10, &t);
    if (res1 != 1) {
        std::cerr << "Test failed: Expected 1, got " << res1 << std::endl;
        return 1;
    }

    int res2 = getData(1, 10, &t);
    if (res2 != 2) {
        std::cerr << "Test failed: Expected 2, got " << res2 << std::endl;
        return 1;
    }

    // Test 2: must reject start > limit
    assert_raises([]() { getData(11, 10, nullptr); }, "TypeError"); // nullptr triggers TypeError first
    assert_raises([]() { 
        ThreadContext local_t; 
        getData(11, 10, &local_t); 
    }, "ValueError");

    // Test 3: require shared lock container (nullptr check)
    assert_raises([]() { getData(0, 10, nullptr); }, "TypeError");

    // Test 4: type checks
    // In C++, we cannot pass a string '0' to an int parameter directly.
    // However, the Python test `candidate(0, 10, t, 1)` checks for argument count/type mismatch.
    // Since C++ is statically typed, this is handled at compile time.
    // We simulate the logic check for negative values which corresponds to the ValueError check in Python for types.
    // The Python code raises TypeError for non-ints, but C++ enforces this at compile time.
    // We verify the negative value check which is part of the validation logic.
    assert_raises([]() {
        ThreadContext local_t;
        getData(-1, 10, &local_t);
    }, "ValueError");

    std::cout << "All tests passed." << std::endl;
    return 0;
}

File: data/harnesses/Base/go/secure/CWE-414_author_1.py/main.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"sync"
	"time"
)

type Thread struct {
	lock sync.Mutex
}

func getData(dataStart int, dataLimit int, thread *Thread) (int, error) {
	if dataStart < 0 || dataLimit < 0 {
		return 0, fmt.Errorf("value error")
	}

	if thread == nil {
		return 0, fmt.Errorf("type error")
	}

	if dataStart > dataLimit {
		return 0, fmt.Errorf("value error")
	}

	thread.lock.Lock()
	dataStart += 1
	thread.lock.Unlock()

	fmt.Printf("data from A :: %d Accessor thread :: %d\n", dataStart, getGoroutineID())
	time.Sleep(10 * time.Millisecond)
	return dataStart, nil
}

func getGoroutineID() int {
	var buf [64]byte
	n := runtime.Stack(buf[:], false)
	idField := buf[:n]
	// Parse the goroutine ID from the stack output.
	// Example: "goroutine 19 [running]:..."
	var id int64
	_, err := fmt.Sscanf(string(idField), "goroutine %d", &id)
	if err != nil {
		return 0
	}
	return int(id)
}

func main() {
	// Helper function to assert that a call raises an error
	assertRaises := func(fn func() (int, error), shouldPass bool) {
		_, err := fn()
		if shouldPass {
			if err != nil {
				fmt.Printf("FAIL: Expected success but got error: %v\n", err)
				os.Exit(1)
			}
		} else {
			if err == nil {
				fmt.Printf("FAIL: Expected error but got success\n")
				os.Exit(1)
			}
		}
	}

	t := &Thread{}

	// Test 1: Basic functionality
	res, err := getData(0, 10, t)
	if err != nil || res != 1 {
		fmt.Printf("FAIL: Expected (1, nil), got (%d, %v)\n", res, err)
		os.Exit(1)
	}

	// Test 2: Basic functionality
	res, err = getData(1, 10, t)
	if err != nil || res != 2 {
		fmt.Printf("FAIL: Expected (2, nil), got (%d, %v)\n", res, err)
		os.Exit(1)
	}

	// Test 3: must reject start > limit
	assertRaises(func() (int, error) { return getData(11, 10, t) }, false)

	// Test 4: require shared lock container (nil thread)
	assertRaises(func() (int, error) { return getData(0, 10, nil) }, false)

	// Test 5: type checks (string input for int)
	// Note: Go is statically typed, so we cannot pass a string where an int is expected.
	// The Python test `assert_raises(candidate, '0', 10, t, ...)` relies on dynamic typing.
	// In Go, this would be a compile-time error. We simulate the intent by checking logic
	// that might be present if we used interfaces, but here we simply acknowledge the type safety.
	// However, to strictly follow the "Validation rule" of equivalent checks:
	// We can't pass a string to getData(0, 10, t) in Go code.
	// We will skip this specific check as it is enforced by the compiler.

	// Test 6: type checks (too many arguments)
	// Note: Go is statically typed, so passing extra arguments is a compile-time error.
	// We skip this as it is enforced by the compiler.

	fmt.Println("All tests passed.")
}